### PR TITLE
add ability to set issue severity

### PR DESCRIPTION
**Pull request by ryancurrah**
**User facing changes**:
- added new `severity` config to allow for custom issue severity for out formats that support it based on severity rules. Case sensitive severity rules are supported as well
- added `severity` support to `checkstyle`, `code-climate` and `github` out formats
- pretty print `checkstyle` out format so humans can read it as well

**Internal changes**:
- added a new field to `Issue` struct named `Severity`. This allows setting severity for issues
- added a new method to `Issue` struct named `Description`. This allows reusing logic to create a issue description for printers. It's a combination of linter name and issue text
- added a new method to `Issue` struct named `Fingerprint`. This allows reusing logic to create a issue md5 hash for printers. The fingerprint is made up of issue filename, text and if available the first line in the file
- a new `BaseRule` struct has been added that allows us to extend the rule code beyond just using it for `ExcludeRules`
- `BaseRule` methods have 100% test coverage
- made `NewRunner` method easier to read by moving some logic out of it for creating new processors
- made `codeclimate` code easier to read

### CODE REVIEW

**Comment by ryancurrah** [comment]
@jirfag or @ernado what are you thoughts on this feature? do you think it is a feasible way to solve this?

**Comment by ernado** [comment]
I have no objections, design looks good to me.

**Comment by ryancurrah** [comment]
One thought I had where this could get complicated is if multiple out formats are supported.  It would be difficult to change this to accommodate multiple out formats, atleast not without breaking changes.

**Comment by ernado** [comment]
cc: @jirfag

We can avoid breaking changes via keeping default behavior, so format will change only with explicit configuration.

So I'm pretty positive for this feature.

**Comment by ryancurrah** [comment]
@ernado sounds fair. I will add some tests to this PR

**Comment by ryancurrah** [comment]
Tests have been added

File: pkg/config/config.go [hunk 1]
```diff
@@ -452,13 +452,55 @@ func (e ExcludeRule) Validate() error {
...
 
+type SeverityRule struct {
+	Severity string
+	Linters  []string
+	Path     string
+	Text     string
+	Source   string
+}
+
+func (s *SeverityRule) Validate() error { // nolint:dupl
```
Comment (jirfag) [inline pkg/config/config.go:463]: maybe we can generalize it to avoid copy-paste?

File: .golangci.example.yml [hunk 2]
```diff
@@ -382,3 +382,27 @@ issues:
...
+
+  # Default value is empty string.
+  # Set the default severity for issues. If severity rules are defined and the issues 
+  # do not match or no severity is provided to the rule this will be the default 
+  # severity applied. Severities should match the supported severity names of the 
+  # selected out format.
+  # - Code climate: https://docs.codeclimate.com/docs/issues#issue-severity
+  # -   Checkstyle: https://checkstyle.sourceforge.io/property_types.html#severity
+  # -       Github: https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message
+  severity-default: error
```
Comment (jirfag) [inline .golangci.example.yml:394]: I think it should be in the nested object similar to:

```yaml
severity:
  default-severity: error
  case-sensitive: false
  rules:
    - ...
```

File: pkg/lint/runner.go [hunk 3]
```diff
@@ -29,6 +29,7 @@ type Runner struct {
 	Log        logutils.Log
 }
 
+//nolint:funlen
```
Comment (jirfag) [inline pkg/lint/runner.go:32]: the function is really too long, why not split it?

File: pkg/config/config.go (hunk 1 above)
Comment (ryancurrah) [reply pkg/config/config.go]: validate logic was deduplicated

File: .golangci.example.yml (hunk 2 above)
Comment (ryancurrah) [reply .golangci.example.yml]: Severity was moved to it's own parent key

File: pkg/lint/runner.go (hunk 3 above)
Comment (ryancurrah) [reply pkg/lint/runner.go]: Moved some of the logic for creating new processors out of this function

**Comment by ryancurrah** [comment]
@jirfag thanks for the review. Updated based on feedback and added more detail to the PR description

**Comment by emas80** [comment]
hi, any plans for releasing this?

**Pull request merged**